src/sema/types.rs

```diff
@@ -203,27 +203,44 @@ fn find_struct_recursion(struct_no: usize, structs_visited: &mut Vec<usize>, ns:
     let mut types_seen: HashSet<usize> = HashSet::new();
 
     for (field_no, field) in def.fields.iter().enumerate() {
-        if let Type::Struct(StructType::UserDefined(field_struct_no)) = field.ty {
-            if types_seen.contains(&field_struct_no) {
-                continue;
+        let field_struct_no = match &field.ty {
+            Type::Struct(StructType::UserDefined(n)) => *n,
+            Type::Mapping(_, v) => {
+                if let Type::Struct(StructType::UserDefined(n)) = **v {
+                    n
+                } else {
+                    continue;
+                }
             }
+            Type::Array(v, _) => {
```

Comment:
I think this is only true for arrays of fixed length (outer dimension).

```
struct S {
      S[] a;
      int b'
}
```
Is permitted (although a little weird).

Just like you can do with rust:

```rust
struct Expression {
     expr: Vec<Expression>,
}
```
but not 
```rust
struct Expression {
     expr: [Expression; 2],
}
```